File: billiards/app/models/journal.py
"""Журнал событий клуба."""

import enum
from datetime import datetime

from sqlalchemy import DateTime, Enum, ForeignKey, String
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base
from app.timeutils import utcnow


class JournalEvent(enum.StrEnum):
    TABLE_CREATED = "table_created"
    TARIFF_CREATED = "tariff_created"
    SESSION_OPENED = "session_opened"
    SESSION_CLOSED = "session_closed"
    LIGHT_ON = "light_on"
    LIGHT_OFF = "light_off"


class JournalEntry(Base):
    __tablename__ = "journal_entries"

    id: Mapped[int] = mapped_column(primary_key=True)
    event: Mapped[JournalEvent] = mapped_column(
        Enum(JournalEvent, values_callable=lambda e: [m.value for m in e]),
        nullable=False,
    )
    message: Mapped[str] = mapped_column(String(500), nullable=False)
    table_id: Mapped[int | None] = mapped_column(
        ForeignKey("tables.id", ondelete="SET NULL"), nullable=True
    )
    session_id: Mapped[int | None] = mapped_column(
        ForeignKey("table_sessions.id", ondelete="SET NULL"), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=utcnow, nullable=False, index=True
    )

File: billiards-node/src/services/journal.js
// Запись событий в журнал. Единственная точка записи: все сервисы
// фиксируют события через logEvent, поэтому формат записей единообразен.

import { utcNow } from "../db.js";

export const JournalEvent = Object.freeze({
  TABLE_CREATED: "table_created",
  TARIFF_CREATED: "tariff_created",
  SESSION_OPENED: "session_opened",
  SESSION_CLOSED: "session_closed",
  LIGHT_ON: "light_on",
  LIGHT_OFF: "light_off",
});

/**
 * Добавляет запись в журнал.
 * @param {import("better-sqlite3").Database} db
 * @param {string} event значение из JournalEvent
 * @param {string} message человекочитаемое описание
 * @param {{tableId?: number, sessionId?: number}} [refs]
 */
export function logEvent(db, event, message, refs = {}) {
  db.prepare(
    `INSERT INTO journal_entries (event, message, table_id, session_id, created_at)
     VALUES (?, ?, ?, ?, ?)`
  ).run(event, message, refs.tableId ?? null, refs.sessionId ?? null, utcNow());
}

/**
 * Последние записи журнала, новые сверху.
 * @param {import("better-sqlite3").Database} db
 * @param {number} [limit]
 */
export function listJournal(db, limit = 200) {
  return db
    .prepare(
      `SELECT id, event, message, table_id, session_id, created_at
       FROM journal_entries
       ORDER BY created_at DESC, id DESC
       LIMIT ?`
    )
    .all(limit);
}
